### ISSUE TITLE
Question regarding TB2K

### ISSUE THREAD

**Issue by pyscripter** [SilverpointDev/sptbxlib#40]
Not an issue but...

In TB2Item.pas and in method TTBView.DoUpdatePositions the last two lines are:
```
  if FWindow.HandleAllocated then
    DoubleBufferedRepaint(FWindow.Handle);

```
In my version I had replaced these lines with:
```
  if FWindow.HandleAllocated then begin
    if (Owner is TWinControl) and TWinControl(Owner).DoubleBuffered then
      InvalidateRect(FWindow.Handle, nil, True)
    else
      DoubleBufferedRepaint(FWindow.Handle);
  end;
```

I remember that introducing this change resulted in a huge reduction in flicker  while resizing with older versions of Sptbxlib.   To be frank with the current state of Sptbxlib I do not see much improvement.
*Questions:*
- Are there any negative side-effects of the above change?
- Do the recent changes in Sptbxlib make the above redundant?

**Comment by pyscripter** [comment]
I have reproduced the problem I was having.  In the form below the toolbar icons in File Explorer are flashing (flickering) as I resize in a very noticeable and annoying way.  (but not the main toolbar icons).
![capture](https://user-images.githubusercontent.com/1311616/49037504-13542b00-f1c3-11e8-9342-a016c82fa8fa.JPG).

This is because VCL, as you resize the form, realigns and invalidates the nested panels. This has to do with Vcl Styles using non standard border size.  Flicker does not happen if you disable the styling of the main form border.   TTBView.DoUpdatePositions instead of invalidating forces the repaint of the toolbar buttons resulting in flickering.   

Unfortunately, I am still getting the flicker even with my change applied.    This is because in the current version, WM_ERASEBACKGROUND repaints the background of the toolbars.  Using my fork I do not get that.

**Comment by pyscripter** [comment]
To sum-up, to avoid this kind of flicker two things are needed.
1. The above change in TTBView.DoUpdatePositions and
2. Avoiding the repainting of the background in the handling of WM_ERASEBACKGROUND

**Comment by SilverpointDev** [comment]
Can't reproduce, can you post a simple demo?

**Comment by pyscripter** [comment]
[Flicker.zip](https://github.com/SilverpointDev/sptbxlib/files/2706412/Flicker.zip)

Here is a simple project reproducing the problem.  To compile it you need Delphi Rio.  If you do not have access to Delphi Rio, please let me know and I will produce another version using Delphi Tokyo.
The project uses my Vcl.Styles.DPIAware unit for using Vcl styles in system DPI Aware applications (included).  The Vcl.Styles.DPIAware unit uses the delphi-detours-library also included.

Run the project, resize and notice the flicker of the SpTBXItems on the toolbars located inside the left-side panels.

Observations:
- SpTBXToolbar inside TPanel -> flicker
- Vcl Toolbar insdide TPanel -> no flicker
- SpTBXToolbar inside SpTBXPanel or SpTBXTabControl -> no flicker
- SpTBXToolbar inside TPageControl -> flicker

So you may ask why don't I use exclusively SpTBX panels and SpTBXTabControls to avoid flicker.  My issue is that I am using the JvDocking library and the toolbars are placed on TPanel/TPageControl like components created on the fly.

**Comment by SilverpointDev** [comment]
Try setting DoubleBuffered on the Docks, in your example it should be:
```pascal
SpTBXDock2.DoubleBuffered := True;
SpTBXDock3.DoubleBuffered := True;
```

**Comment by pyscripter** [comment]
Yes, that resolves the flicker.  Setting the double-buffer property of the Toolbar instead of the dock also stops the flicker.  By the way, when you set the dock doublebuffer property to True, this also turns on the doublebuffer property of the contained toolbars.  This because the ParentDoubleProperty is True by default.

*However now if you drag the toolbar the background turns black, i.e. it is not updated.*

By the way why not making the DoubleBuffer/ParentDoubleBuffer properties of Docks and ToolBars published?

**Comment by SilverpointDev** [comment]
I forgot to change a line in TB2Dock, change this in TTBCustomDockableWindow.WMMove:
```
SendMessage(W, WM_ERASEBKGND, WPARAM(BmpDC), 0);
```
To:
```
SendMessage(W, WM_ERASEBKGND, WPARAM(BmpDC), LPARAM(BmpDC)); // Robert: Pass BmpDC on LParam to support DoubleBuffered property
```

**Comment by pyscripter** [comment]
Great!  This resolves the issue.
So what is the recommended way?
a)  Setting the DoubleBuffered property of Docks to true?
b)  Setting the DoubleBuffered property of Toolbars to true?
c)  Both?

Doesn't it make sense to publish the DoubleBuffered property?
Or else/in addition set it to true for Docks/Toolbars in their constructor?

**Comment by pyscripter** [comment]
Could you please provide some advice on the above so that the issue can be closed.

**Comment by SilverpointDev** [comment]
Set DoubleBeffered on the Docks
